src/Microsoft.ML.StaticPipe/KMeansStatic.cs

```diff
@@ -30,19 +31,62 @@ public static class KMeansClusteringExtensions
            Action<KMeansModelParameters> onFit = null)
         {
             Contracts.CheckValue(features, nameof(features));
             Contracts.CheckValueOrNull(weights);
             Contracts.CheckParam(clustersCount > 1, nameof(clustersCount), "If provided, must be greater than 1.");
             Contracts.CheckValueOrNull(onFit);
-            Contracts.CheckValueOrNull(advancedSettings);
 
             var rec = new TrainerEstimatorReconciler.Clustering(
             (env, featuresName, weightsName) =>
             {
-                var trainer = new KMeansPlusPlusTrainer(env, featuresName, clustersCount, weightsName, advancedSettings);
+                var options = new KMeansPlusPlusTrainer.Options
+                {
+                    FeatureColumn = featuresName,
+                    K = clustersCount,
+                    WeightColumn = weightsName != null ? Optional<string>.Explicit(weightsName) : Optional<string>.Implicit(DefaultColumnNames.Weight)
+                };
+
+                var trainer = new KMeansPlusPlusTrainer(env, options);
+
+                if (onFit != null)
+                    return trainer.WithOnFitDelegate(trans => onFit(trans.Model));
+                else
+                    return trainer;
+            }, features, weights);
+
+            return rec.Output;
+        }
+
+        /// <summary>
+        /// KMeans <see cref="ClusteringContext"/> extension method.
+        /// </summary>
+        /// <param name="ctx">The regression context trainer object.</param>
+        /// <param name="features">The features, or independent variables.</param>
+        /// <param name="weights">The optional example weights.</param>
+        /// <param name="options">Algorithm advanced settings.</param>
+        /// <param name="onFit">A delegate that is called every time the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}.Fit(DataView{TInShape})"/> method is called on the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}"/> instance created out of this. This delegate will receive
+        /// the linear model that was trained.  Note that this action cannot change the result in any way; it is only a way for the caller to
+        /// be informed about what was learnt.</param>
+        /// <returns>The predicted output.</returns>
+        public static (Vector<float> score, Key<uint> predictedLabel) KMeans(this ClusteringContext.ClusteringTrainers ctx,
+           Vector<float> features, Scalar<float> weights = null,
+           KMeansPlusPlusTrainer.Options options = null,
+           Action<KMeansModelParameters> onFit = null)
+        {
+            Contracts.CheckValueOrNull(onFit);
+            Contracts.CheckValueOrNull(options);
+
+            var rec = new TrainerEstimatorReconciler.Clustering(
+            (env, featuresName, weightsName) =>
+            {
+                options.FeatureColumn = featuresName;
```

Comment:
>options [](start = 16, length = 7)

you need to check are they null, and create new one if they are.

Reply:
that's what happen when you are in a hurry to press publish.. 

---
In reply to: [248119409](https://github.com/dotnet/machinelearning/pull/2158#discussion_r248119409) [](ancestors = 248119409)